File: domain/ports/incoming/service/src/main/java/uk/ratracejoe/sdq/service/SdqService.java
package uk.ratracejoe.sdq.service;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import uk.ratracejoe.sdq.exception.SdqException;
import uk.ratracejoe.sdq.model.Assessor;
import uk.ratracejoe.sdq.model.ReportingPeriod;
import uk.ratracejoe.sdq.model.demographics.DemographicFilter;
import uk.ratracejoe.sdq.model.sdq.SdqProgressSummary;
import uk.ratracejoe.sdq.model.sdq.SdqSubmission;
import uk.ratracejoe.sdq.model.sdq.SdqSubmissionSummary;

public interface SdqService {

  void recordResponse(SdqSubmission sdq) throws SdqException;

  SdqSubmission getSubmission(UUID periodId, Assessor assessor);

  SdqSubmissionSummary getSubmissionSummary(UUID periodId, Assessor assessor);

  List<ReportingPeriod> getReportingPeriods(UUID clientId);

  List<SdqProgressSummary> querySdqProgress(
      List<Assessor> assessors, List<DemographicFilter> filters, LocalDate from, LocalDate to);

  List<SdqSubmissionSummary> querySdqSummaries(
      List<Assessor> assessors, List<DemographicFilter> filters, LocalDate from, LocalDate to);

  SdqProgressSummary getSdqProgressForClient(UUID clientId, Assessor assessor);
}

File: sdq-api-rust/db/src/sdq_service.rs
use async_trait::async_trait;
use chrono::NaiveDate;
use sdq_model::{
    Assessor, DemographicFilter, ReportingPeriod, SdqError, SdqProgressSummary, SdqSubmission,
    SdqSubmissionSummary,
};
use sdq_service::sdq::SdqService;
use sqlx::PgPool;
use std::collections::HashMap;
use uuid::Uuid;

pub struct SdqServiceSqlxImpl {
    pub pool: PgPool,
}

impl SdqServiceSqlxImpl {
    pub fn new(pool: PgPool) -> SdqServiceSqlxImpl {
        SdqServiceSqlxImpl { pool }
    }
}

#[async_trait]
impl SdqService for SdqServiceSqlxImpl {
    async fn record_response(&self, sdq: SdqSubmission) -> Result<(), SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_submission(
        &self,
        period_id: &Uuid,
        assessor: &Assessor,
    ) -> Result<SdqSubmission, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_summary(
        &self,
        period_id: &Uuid,
        assessor: &Assessor,
    ) -> Result<SdqSubmissionSummary, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn get_reporting_periods(
        &self,
        client_id: &Uuid,
    ) -> Result<Vec<ReportingPeriod>, SdqError> {
        Err(SdqError::NotImplemented)
    }
    async fn query_sdq_progress(
        &self,
        assessors: &Vec<Assessor>,
        filters: &Vec<DemographicFilter>,
        from: &NaiveDate,
        to: &NaiveDate,
    ) -> Result<Vec<SdqProgressSummary>, SdqError> {
        Err(SdqError::NotImplemented)
    }

    async fn query_sdq_summaries(
        &self,
        assessors: &Vec<Assessor>,
        filters: &Vec<DemographicFilter>,
        from: &NaiveDate,
        to: &NaiveDate,
    ) -> Result<HashMap<Uuid, HashMap<NaiveDate, Vec<SdqSubmissionSummary>>>, SdqError> {
        Err(SdqError::NotImplemented)
    }

    async fn get_sdq_progress_for_client(
        &self,
        client_id: &Uuid,
        assessor: &Assessor,
    ) -> Result<SdqProgressSummary, SdqError> {
        Err(SdqError::NotImplemented)
    }
}
